[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73436
* #73386
* #73385
* #73233
* __->__ #73120

Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

jerryzh168 committed:
[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler

Summary:
att
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -11,13 +11,27 @@
+    ['running_mean', 'running_var', 'use_input_stats', 'momentum'],
+}
+
+def _is_node(node, modules, func_list, method_list, module_type_list):
```
maybe something like `_is_node_in_list`?

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -11,13 +11,27 @@
 from torch.fx import Node
 import operator
 
+ARGS_TO_SKIP = {
```
maybe `qfun_to_arg_names_to_skip`?

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for functional ops using DefaultNodeQuantizeHandler"


Summary:
This is to align our implementation with https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

This changes the ops in DefaultNodeQuantizeHandler to produce reference patterns: dequantize - fp32_op - quantize
and then adds a lowering pass to lower them to quantized ops.

Most of them are straightforward, except for ConvTranspose1d and ConsTranspose2d, which needs reference module support, and we need to add a `from_reference` classmethod in `nnq.ConvTranspose{N}d`

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34354038](https://our.internmc.facebook.com/intern/diff/D34354038)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34354038).
